File: exams/frmContribution.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace exams
{
    public partial class frmContribution
    {
        public frmContribution()
        {
            InitializeComponent();
            _cboYear.Name = "cboYear";
            _cboTerm.Name = "cboTerm";
            _btnAddExam.Name = "btnAddExam";
            _btnEnterMarks.Name = "btnEnterMarks";
            _btnClear.Name = "btnClear";
            _btnCancel.Name = "btnCancel";
        }

        private void frmContribution_Load(object sender, EventArgs e)
        {
            if (!publicSubsNFunctions.connect())
            {
                Close();
            }
            else
            {
                total_percentage = 0d;
                txtContribution.Clear();
                lstExaminations.Items.Clear();
                grpMultiExaminations.Enabled = true;
                fill_year();
                publicSubsNFunctions.get_term();
                var argcbo = cboClass;
                publicSubsNFunctions.load_class(ref argcbo);
                cboClass = argcbo;
                cboTerm.SelectedItem = publicSubsNFunctions.term;
                txtContribution.Text = 0.ToString();
                cboSelectedTerm.SelectedItem = publicSubsNFunctions.term;
            }
        }

        private void fill_year()
        {
            cboYear.Items.Clear();
            cboSelectedYear.Items.Clear();
            var argcbo = cboYear;
            publicSubsNFunctions.fill_years(ref argcbo);
            cboYear = argcbo;
            var argcbo1 = cboSelectedYear;
            publicSubsNFunctions.fill_years(ref argcbo1);
            cboSelectedYear = argcbo1;
            cboYear.SelectedItem = DateAndTime.Today.Year;
            cboSelectedTerm.SelectedItem = DateAndTime.Today.Year;
        }

        private void btnAddExam_Click(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(isvalid()))
            {
                add_exam();
            }
            else
            {
                Interaction.MsgBox(msg);
            }
        }

        private string getExamContribution()
        {
            string contribution = string.Empty;
            string argq = "select total from examinations where term = '" + cboTerm.SelectedItem.ToString() + "' and year = '" + cboYear.SelectedItem.ToString() + "' and examname = '" + cboExamName.SelectedItem.ToString() + "'";
            if (publicSubsNFunctions.qread(ref argq))
            {
                while (publicSubsNFunctions.dbreader.Read())
                    contribution = Conversions.ToString(publicSubsNFunctions.dbreader["total"]);
            }

            return contribution;
        }

        private double total_percentage = 0d;
        private string msg = string.Empty;

        private object isvalid()
        {
            double buffer = 0d;
            buffer = total_percentage + Conversions.ToDouble(txtContribution.Text);
            if (buffer > 100d)
            {
                msg = "The Examination Contributions Must Not Exceed 100%";
                return false;
            }

            if (!Information.IsNumeric(cboYear.SelectedItem))
            {
                msg = "Invalid Selection For Year!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false)))
            {
                msg = "Invalid Selection for Term!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboExamName.SelectedItem, publicSubsNFunctions.None, false)))
            {
                msg = "Invalid Selection For Exam Name!";
                return false;
            }
            else if (!Information.IsNumeric(txtContribution.Text))
            {
                msg = "The Value For Contribution Is Not A Number";
                return false;
            }
            else if (Conversions.ToDouble(txtContribution.Text) == 0d)
            {
                msg = "Contribution Can Not Be Zero";
                return false;
            }
            else
            {
                return true;
            }
        }

        private object isValidSave()
        {
            if (Conversions.ToBoolean(Operators.OrObject(Operators.ConditionalCompareObjectEqual(cboClass.SelectedItem, null, false), Operators.ConditionalCompareObjectEqual(cboClass.SelectedItem, publicSubsNFunctions.None, false))))
            {
                msg = "Invalid Selection For Class!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, string.Empty, false), cboSortBy.Visible)))
            {
                msg = "Please Choose The Mode To Sort The Result Analysis!";
                return false;
            }
            else
            {
                return true;
            }
        }

        private void add_exam()
        {
            // todo add method to get the exam out of
            if (Conversions.ToBoolean(Operators.AndObject(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboExamName.SelectedItem, publicSubsNFunctions.None, false), Information.IsNumeric(txtContribution.Text)), (txtContribution.Text ?? "") != (string.Empty ?? ""))))
            {
                if (Conversions.ToBoolean(!exists(Conversions.ToString(cboExamName.SelectedItem))))
                {
                    var li = new ListViewItem();
                    li = lstExaminations.Items.Add(cboExamName.SelectedItem);
                    li.SubItems.Add(txtContribution.Text);
                    li.SubItems.Add(cboTerm.SelectedItem);
                    li.SubItems.Add(cboYear.SelectedItem);
                    li.SubItems.Add(getExamContribution());
                    total_percentage += Conversions.ToDouble(txtContribution.Text);
                }
            }
        }

        private object exists(string str)
        {
            int i;
            string selectedTerm = string.Empty;
            var loopTo = lstExaminations.Items.Count - 1;
            for (i = 0; i <= loopTo; i++)
            {
                if (Conversions.ToBoolean(Operators.AndObject(Operators.AndObject((lstExaminations.Items[i].Text ?? "") == (str ?? ""), Operators.ConditionalCompareObjectEqual(lstExaminations.Items[i].SubItems[2].Text, cboYear.SelectedItem, false)), Operators.ConditionalCompareObjectEqual(lstExaminations.Items[i].SubItems[1].Text, cboTerm.SelectedItem, false))))
                {
                    Interaction.MsgBox("The Exam Already Exist In The List");
                    return true;
                }
            }

            return false;
        }

        private void cboTerm_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false), (cboYear.SelectedItem.ToString() ?? "") != (publicSubsNFunctions.None ?? ""))))
            {
                fill_exam();
            }
        }

        private void fill_exam()
        {
            string argq = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject("SELECT ExamName FROM examinations WHERE Term='", cboTerm.SelectedItem), "' AND Year='"), cboYear.SelectedItem), "'"));
            if (publicSubsNFunctions.qread(ref argq))
            {
                cboExamName.Items.Clear();
                cboExamName.Items.Add(publicSubsNFunctions.None);
                while (publicSubsNFunctions.dbreader.Read())
                    cboExamName.Items.Add(publicSubsNFunctions.dbreader["ExamName"]);
                cboExamName.SelectedItem = publicSubsNFunctions.None;
                publicSubsNFunctions.dbreader.Close();
            }
            else
            {
                publicSubsNFunctions.failure("Could Not Load Examinations!");
            }
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            publicSubsNFunctions.search_teachers = false;
            Close();
        }

        private void cboYear_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false), (cboYear.SelectedItem.ToString() ?? "") != (publicSubsNFunctions.None ?? ""))))
            {
                fill_exam();
            }
        }

        private void btnEnterMarks_Click(object sender, EventArgs e)
        {
            if (lstExaminations.Items.Count == 0)
            {
                Interaction.MsgBox("There Are No Exams To Analyze");
            }
            else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboClass.SelectedItem, null, false)))
            {
                Interaction.MsgBox("Please Select The Class");
            }
            else if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, string.Empty, false), cboSortBy.Visible)))
            {
                Interaction.MsgBox("Please Choose The Mode To Sort The Result Analysis!");
            }
            else
            {
                if (rdSubjectBased.Checked == true)
                {
                    publicSubsNFunctions.gradingType = "SubjectBased";
                }
                else
                {
                    publicSubsNFunctions.gradingType = "ClassBased";
                }

                publicSubsNFunctions.selectedTerm = cboTerm.SelectedItem.ToString();
                publicSubsNFunctions.selectedClass = cboClass.SelectedItem.ToString();
                publicSubsNFunctions.selectedYear = cboSelectedYear.SelectedItem.ToString();
                publicSubsNFunctions.sortGradesBy = cboSortBy.SelectedItem.ToString();
                publicSubsNFunctions.tm = Conversions.ToString(cboTerm.SelectedItem);
                publicSubsNFunctions.yr = Conversions.ToInteger(cboYear.SelectedItem);
                publicSubsNFunctions.examListMain = new List<Tuple<string, string, string, string, string>>();
                for (int i = 0, loopTo = lstExaminations.Items.Count - 1; i <= loopTo; i++)
                    publicSubsNFunctions.examListMain.Add(new Tuple<string, string, string, string, string>(lstExaminations.Items[i].Text, lstExaminations.Items[i].SubItems[1].Text, lstExaminations.Items[i].SubItems[2].Text, lstExaminations.Items[i].SubItems[3].Text, lstExaminations.Items[i].SubItems[4].Text));
                publicSubsNFunctions.subjectColumns = new List<string>();
                string argq = "SHOW COLUMNS FROM subjects_done;";
                publicSubsNFunctions.qread(ref argq);
                string columns = string.Empty;
                while (publicSubsNFunctions.dbreader.Read())
                {
                    columns = Conversions.ToString(publicSubsNFunctions.dbreader["Field"]);
                    if (columns != "ADMNo")
                    {
                        publicSubsNFunctions.subjectColumns.Add(columns);
                    }
                }

                // //////////////////////////////////////////////////////////////////////

                publicSubsNFunctions.table = "exam_results";
                publicSubsNFunctions.search_teachers = true;
                publicSubsNFunctions.tm = Conversions.ToString(cboTerm.SelectedItem);
                publicSubsNFunctions.yr = Conversions.ToInteger(cboYear.SelectedItem);
                if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Total Marks", false)))
                {
                    publicSubsNFunctions.sortby = "Total";
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Mean Marks", false)))
                {
                    publicSubsNFunctions.sortby = "MM";
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Mean Points", false)))
                {
                    publicSubsNFunctions.sortby = "MP";
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Total Points", false)))
                {
                    publicSubsNFunctions.sortby = "TP";
                }

                // /////////////////////////////////////////////////////////////////////

                Close();
                // total be converted to devexpress
                var computeResultForm = new frmComputeResults();
                computeResultForm.MdiParent = My.MyProject.Forms.frmMainForm;
                computeResultForm.Show();
            }
        }

        private void btnClear_Click(object sender, EventArgs e)
        {
            clear();
            total_percentage = 0d;
        }

        private void clear()
        {
            cboTerm.SelectedItem = publicSubsNFunctions.None;
            cboExamName.SelectedItem = publicSubsNFunctions.None;
            cboClass.SelectedItem = publicSubsNFunctions.None;
            lstExaminations.Items.Clear();
            txtContribution.Text = 0.ToString();
        }
    }
}

File: exams/frmContribution.vb
Public Class frmContribution

    Private Sub frmContribution_Load(sender As Object, e As EventArgs) Handles MyBase.Load

        If Not connect() Then
            Me.Close()
        Else
            total_percentage = 0
            txtContribution.Clear()
            lstExaminations.Items.Clear()
            grpMultiExaminations.Enabled = True
            fill_year()
            get_term()
            load_class(cboClass)
            cboTerm.SelectedItem = term
            txtContribution.Text = 0
            cboSelectedTerm.SelectedItem = term
        End If
    End Sub

    Private Sub fill_year()
        cboYear.Items.Clear()
        cboSelectedYear.Items.Clear()
        fill_years(cboYear)
        fill_years(cboSelectedYear)
        cboYear.SelectedItem = Today.Year
        cboSelectedTerm.SelectedItem = Today.Year
    End Sub

    Private Sub btnAddExam_Click(sender As Object, e As EventArgs) Handles btnAddExam.Click
        If isvalid() Then
            add_exam()
        Else
            MsgBox(msg)
        End If

    End Sub

    Private Function getExamContribution() As String
        Dim contribution As String = String.Empty
        If qread("select total from examinations where term = '" + cboTerm.SelectedItem.ToString + "' and year = '" + cboYear.SelectedItem.ToString + "' and examname = '" + cboExamName.SelectedItem.ToString + "'") Then
            While dbreader.Read
                contribution = dbreader("total")
            End While
        End If
        Return contribution
    End Function

    Dim total_percentage As Double = 0
    Dim msg As String = String.Empty
    Private Function isvalid()

        Dim buffer As Double = 0
        buffer = total_percentage + txtContribution.Text
        If buffer > 100 Then
            msg = "The Examination Contributions Must Not Exceed 100%"
            Return False
        End If

        If Not IsNumeric(cboYear.SelectedItem) Then
            msg = "Invalid Selection For Year!"
            Return False
        ElseIf cboTerm.SelectedItem = None Then
            msg = "Invalid Selection for Term!"
            Return False
        ElseIf cboExamName.SelectedItem = None Then
            msg = "Invalid Selection For Exam Name!"
            Return False
        ElseIf Not IsNumeric(txtContribution.Text) Then
            msg = "The Value For Contribution Is Not A Number"
            Return False
        ElseIf txtContribution.Text = 0 Then
            msg = "Contribution Can Not Be Zero"
            Return False
        Else
            Return True
        End If

    End Function

    Private Function isValidSave()
        If cboClass.SelectedItem = Nothing Or cboClass.SelectedItem = None Then
            msg = "Invalid Selection For Class!"
            Return False
        ElseIf cboSortBy.SelectedItem = String.Empty And cboSortBy.Visible Then
            msg = "Please Choose The Mode To Sort The Result Analysis!"
            Return False
        Else
            Return True
        End If
    End Function

    Private Sub add_exam()
        'todo add method to get the exam out of
        If cboExamName.SelectedItem <> None And IsNumeric(txtContribution.Text) And txtContribution.Text <> String.Empty Then
            If Not exists(cboExamName.SelectedItem) Then
                Dim li As New ListViewItem
                li = lstExaminations.Items.Add(cboExamName.SelectedItem)
                li.SubItems.Add(txtContribution.Text)
                li.SubItems.Add(cboTerm.SelectedItem)
                li.SubItems.Add(cboYear.SelectedItem)
                li.SubItems.Add(getExamContribution())
                total_percentage += txtContribution.Text
            End If
        End If
    End Sub

    Private Function exists(ByVal str As String)
        Dim i As Integer
        Dim selectedTerm As String = String.Empty

        For i = 0 To lstExaminations.Items.Count - 1
            If lstExaminations.Items.Item(i).Text = str And lstExaminations.Items.Item(i).SubItems(2).Text = cboYear.SelectedItem And lstExaminations.Items.Item(i).SubItems(1).Text = cboTerm.SelectedItem Then
                MsgBox("The Exam Already Exist In The List")
                Return True
            End If
        Next
        Return False
    End Function

    Private Sub cboTerm_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboTerm.SelectedIndexChanged
        If cboTerm.SelectedItem <> None And cboYear.SelectedItem.ToString <> None Then
            fill_exam()
        End If
    End Sub

    Private Sub fill_exam()
        If qread("SELECT ExamName FROM examinations WHERE Term='" & cboTerm.SelectedItem & "' AND Year='" & cboYear.SelectedItem & "'") Then
            cboExamName.Items.Clear()
            cboExamName.Items.Add(None)
            While dbreader.Read
                cboExamName.Items.Add(dbreader("ExamName"))
            End While
            cboExamName.SelectedItem = None
            dbreader.Close()
        Else
            failure("Could Not Load Examinations!")
        End If
    End Sub

    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        search_teachers = False
        Me.Close()
    End Sub

    Private Sub cboYear_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboYear.SelectedIndexChanged
        If cboTerm.SelectedItem <> None And cboYear.SelectedItem.ToString <> None Then
            fill_exam()
        End If
    End Sub

    Private Sub btnEnterMarks_Click(sender As Object, e As EventArgs) Handles btnEnterMarks.Click
        If lstExaminations.Items.Count = 0 Then
            MsgBox("There Are No Exams To Analyze")
        ElseIf cboClass.SelectedItem = Nothing Then
            MsgBox("Please Select The Class")
        ElseIf cboSortBy.SelectedItem = String.Empty And cboSortBy.Visible Then
            MsgBox("Please Choose The Mode To Sort The Result Analysis!")
        Else

            If rdSubjectBased.Checked = True Then
                gradingType = "SubjectBased"
            Else
                gradingType = "ClassBased"
            End If
            selectedTerm = cboTerm.SelectedItem.ToString()
            selectedClass = cboClass.SelectedItem.ToString()
            selectedYear = cboSelectedYear.SelectedItem.ToString
            sortGradesBy = cboSortBy.SelectedItem.ToString()

            tm = cboTerm.SelectedItem
            yr = cboYear.SelectedItem

            examListMain = New List(Of Tuple(Of String, String, String, String, String))

            For i = 0 To lstExaminations.Items.Count - 1
                examListMain.Add(New Tuple(Of String, String, String, String, String)(lstExaminations.Items.Item(i).Text, lstExaminations.Items.Item(i).SubItems(1).Text, lstExaminations.Items.Item(i).SubItems(2).Text, lstExaminations.Items.Item(i).SubItems(3).Text, lstExaminations.Items.Item(i).SubItems(4).Text))
            Next

            subjectColumns = New List(Of String)
            qread("SHOW COLUMNS FROM subjects_done;")
            Dim columns As String = String.Empty
            While dbreader.Read
                columns = dbreader("Field")

                If columns <> "ADMNo" Then
                    subjectColumns.Add(columns)
                End If
            End While

            '//////////////////////////////////////////////////////////////////////

            table = "exam_results"
            search_teachers = True
            tm = cboTerm.SelectedItem
            yr = cboYear.SelectedItem

            If cboSortBy.SelectedItem = "Total Marks" Then
                sortby = "Total"
            ElseIf cboSortBy.SelectedItem = "Mean Marks" Then
                sortby = "MM"
            ElseIf cboSortBy.SelectedItem = "Mean Points" Then
                sortby = "MP"
            ElseIf cboSortBy.SelectedItem = "Total Points" Then
                sortby = "TP"
            End If

            '/////////////////////////////////////////////////////////////////////

            Me.Close()
            'total be converted to devexpress
            Dim computeResultForm As New frmComputeResults
            computeResultForm.MdiParent = frmMainForm
            computeResultForm.Show()

        End If
    End Sub

    Private Sub btnClear_Click(sender As Object, e As EventArgs) Handles btnClear.Click
        clear()
        total_percentage = 0
    End Sub

    Private Sub clear()
        cboTerm.SelectedItem = None
        cboExamName.SelectedItem = None
        cboClass.SelectedItem = None
        lstExaminations.Items.Clear()
        txtContribution.Text = 0
    End Sub


End Class